contrib/native/client/src/clientlib/drillClientImpl.cpp

```diff
@@ -65,108 +66,70 @@ struct ToRpcType: public std::unary_function<google::protobuf::int32, exec::user
-        if(err){
-            return handleConnError(CONN_ZOOKEEPER_ERROR, getMessage(ERR_CONN_ZOOKEEPER, zook.getError().c_str()));
-        }
-        zook.close();
-        m_bIsDirectConnection=true;
-    }else if(protocol == "local"){
-        boost::lock_guard<boost::mutex> lock(m_dcMutex);//strtok is not reentrant
-        char tempStr[MAX_CONNECT_STR+1];
-        strncpy(tempStr, hostPortStr.c_str(), MAX_CONNECT_STR); tempStr[MAX_CONNECT_STR]=0;
-        host=strtok(tempStr, ":");
-        port=strtok(NULL, "");
-        m_bIsDirectConnection=false;
-    }else{
-        return handleConnError(CONN_INVALID_INPUT, getMessage(ERR_CONN_UNKPROTO, protocol.c_str()));
-    }
-    DRILL_MT_LOG(DRILL_LOG(LOG_TRACE) << "Connecting to endpoint: " << host << ":" << port << std::endl;)
-    std::string serviceHost;
-    for (size_t i = 0; i < props->size(); i++) {
-        if (props->keyAt(i) == USERPROP_SERVICE_HOST) {
-            serviceHost = props->valueAt(i);
-        }
+    connectionStatus_t ret = CONN_SUCCESS;
+    m_pChannelContext = ChannelContextFactory::getChannelContext(type, props);
+    m_pChannel= ChannelFactory::getChannel(type, m_io_service, connStr);
+    ret=m_pChannel->init(m_pChannelContext);
+    if(ret!=CONN_SUCCESS){
+        handleConnError(m_pChannel->getError());
+        return ret;
     }
-    if (serviceHost.empty()) {
-        props->setProperty(USERPROP_SERVICE_HOST, host);
+    ret= m_pChannel->connect();
+    if(ret!=CONN_SUCCESS){
+        handleConnError(m_pChannel->getError());
+        return ret;
     }
-    connectionStatus_t ret = this->connect(host.c_str(), port.c_str());
+    props->setProperty(USERPROP_SERVICE_HOST, m_pChannel->getEndpoint()->getHost());
     return ret;
 }
 
-connectionStatus_t DrillClientImpl::connect(const char* host, const char* port){
-    using boost::asio::ip::tcp;
-    tcp::endpoint endpoint;
-    try{
-        tcp::resolver resolver(m_io_service);
-        tcp::resolver::query query(tcp::v4(), host, port);
-        tcp::resolver::iterator iter = resolver.resolve(query);
-        tcp::resolver::iterator end;
-        while (iter != end){
-            endpoint = *iter++;
-            DRILL_MT_LOG(DRILL_LOG(LOG_TRACE) << endpoint << std::endl;)
-        }
-        boost::system::error_code ec;
-        m_socket.connect(endpoint, ec);
-        if(ec){
-            return handleConnError(CONN_FAILURE, getMessage(ERR_CONN_FAILURE, host, port, ec.message().c_str()));
-        }
-
-    }catch(const std::exception & e){
-        // Handle case when the hostname cannot be resolved. "resolve" is hard-coded in boost asio resolver.resolve
-        if (!strcmp(e.what(), "resolve")) {
-            return handleConnError(CONN_HOSTNAME_RESOLUTION_ERROR, getMessage(ERR_CONN_EXCEPT, e.what()));
+connectionStatus_t DrillClientImpl::connect(const char* host, const char* port, DrillUserProperties* props){
+    if (this->m_bIsConnected || (this->m_pChannelContext!=NULL && this->m_pChannel!=NULL)) {
```

Comment:
I think the second condition is not sufficient to say if there is already a valid connection. Instead there should be `isConnected()` method in channel which will `return m_state == CHANNEL_CONNECTED`

Also why we need both `this->m_bIsConnected` and second condition ?